infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/multiuser/oauth/IdentityProviderConfigBuilder.java

```diff
@@ -0,0 +1,171 @@
+import java.util.Optional;
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+import org.eclipse.che.api.core.BadRequestException;
+import org.eclipse.che.api.core.UnauthorizedException;
+import org.eclipse.che.api.workspace.server.WorkspaceRuntimes;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.api.workspace.server.spi.RuntimeContext;
+import org.eclipse.che.commons.annotation.Nullable;
+import org.eclipse.che.commons.env.EnvironmentContext;
+import org.eclipse.che.commons.subject.Subject;
+import org.eclipse.che.multiuser.keycloak.server.KeycloakServiceClient;
+import org.eclipse.che.multiuser.keycloak.server.KeycloakSettings;
+import org.eclipse.che.multiuser.keycloak.shared.dto.KeycloakTokenResponse;
+import org.eclipse.che.workspace.infrastructure.openshift.ConfigBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class retrieves the OpenShift OAuth token of the current Che user, and injects it the
+ * OpenShift {@link Config} object, so that workspace OpenShift resources will be created under the
+ * OpenShift account of the current Che user.
+ *
+ * <p>The OpenShift OAuth token is retrieved using the OpenShift identity provider configured in the
+ * Keycloak server.
+ *
+ * <p>If the current user is not the user that started the current workspace (for operations such as
+ * idling, che server shutdown, etc ...), then global OpenShift infrastructure credentials defined
+ * in the Che properties will be used.
+ *
+ * @author David Festal
+ */
+@Singleton
+public class IdentityProviderConfigBuilder extends ConfigBuilder {
+
+  private static final Logger LOG = LoggerFactory.getLogger(IdentityProviderConfigBuilder.class);
+
+  private final String oauthIdentityProvider;
+
+  private final KeycloakServiceClient keycloakServiceClient;
+  private final KeycloakSettings keycloakSettings;
+  private final Provider<WorkspaceRuntimes> workspaceRuntimeProvider;
+
+  private String rootUrl;
+
+  @Inject
+  public IdentityProviderConfigBuilder(
+      KeycloakServiceClient keycloakServiceClient,
+      KeycloakSettings keycloakSettings,
+      Provider<WorkspaceRuntimes> workspaceRuntimeProvider,
+      @Nullable @Named("che.infra.openshift.oauth_identity_provider") String oauthIdentityProvider,
+      @Named("che.api") String apiEndpoint) {
+    super();
+    this.keycloakServiceClient = keycloakServiceClient;
+    this.keycloakSettings = keycloakSettings;
+    this.workspaceRuntimeProvider = workspaceRuntimeProvider;
+
+    this.oauthIdentityProvider = oauthIdentityProvider;
+    rootUrl = apiEndpoint;
+    if (rootUrl.endsWith("/")) {
+      rootUrl = rootUrl.substring(0, rootUrl.length() - 1);
+    }
+    if (rootUrl.endsWith("/api")) {
+      rootUrl = rootUrl.substring(0, rootUrl.length() - 4);
+    }
+  }
+
+  /**
+   * Builds the Openshift {@link Config} object based on a default {@link Config} object and an
```

Comment:
Provided config, not default.

Reply:
see the related answer in comment https://github.com/eclipse/che/pull/9577#discussion_r191043430